### PR TITLE
Build frequency KPI for TEngine

### PR DESCRIPTION
**Pull request by aksshriv1**

### CODE REVIEW

File: customapi/src/main/java/com/publicissapient/kpidashboard/apis/jenkins/service/BuildFrequencyServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+			Map<String, List<DataCount>> aggDataMap = new HashMap<>();
+			List<Build> aggBuildList = new ArrayList<>();
+			if (CollectionUtils.isNotEmpty(buildListProjectWise)) {
+				Map<String, List<Build>> buildMapJobWise = buildListProjectWise.stream()
+						.collect(Collectors.groupingBy(Build::getBuildJob, Collectors.toList()));
+				for (Map.Entry<String, List<Build>> entry : buildMapJobWise.entrySet()) {
+					String jobName;
+					List<Build> buildList = entry.getValue();
+					if (StringUtils.isNotEmpty(buildList.get(0).getJobFolder())) {
+						jobName = buildList.get(0).getJobFolder() + CommonConstant.ARROW + trendLineName;
```
Comment (purushottam08) [inline customapi/src/main/java/com/publicissapient/kpidashboard/apis/jenkins/service/BuildFrequencyServiceImpl.java:203]: No need of this here, do it directly in prepareinfoforbuild that way jobname is not needed as argument.

**Review by purushottam08 — APPROVED** [review]

**Review by shunaray — CHANGES_REQUESTED** [review]
Missing:
Coverage
Mongock changes
Fieldmappingdto changes

File: customapi/src/main/java/com/publicissapient/kpidashboard/apis/jenkins/service/BuildFrequencyServiceImpl.java (hunk 1 above)
Comment (aksshriv1) [reply customapi/src/main/java/com/publicissapient/kpidashboard/apis/jenkins/service/BuildFrequencyServiceImpl.java:203]: It is needed here.

File: common/src/main/java/com/publicissapient/kpidashboard/common/model/application/FieldMapping.java [hunk 2]
```diff
@@ -445,6 +445,7 @@ public class FieldMapping extends BasicModel {
 	private String thresholdValueKPI39;
 	private String thresholdValueKPI46;
 	private String thresholdValueKPI8;
+	private String thresholdValueKPI172;
```
Comment (shunaray) [inline common/src/main/java/com/publicissapient/kpidashboard/common/model/application/FieldMapping.java:448]: fieldmapping enum is not present

**Comment by aksshriv1** [comment]
Junit coverage is 85% of the developed feature

File: customapi/src/main/resources/json/mongock/default/kpi_master.json [hunk 3]
```diff
@@ -4020,6 +4020,47 @@
...
+    "kpiId": "kpi172",
+    "kpiName": "Build Frequency",
+    "maxValue": "",
+    "kpiUnit": "",
+    "isDeleted": "False",
+    "defaultOrder": 24,
+    "kpiSource": "Jenkins",
+    "groupId": 1,
+    "thresholdValue": "",
+    "kanban": false,
```
Comment (shunaray) [inline customapi/src/main/resources/json/mongock/default/kpi_master.json:4034]: enter instance thresholdValue as in req, also bg color for threshold is missing

File: customapi/src/main/resources/json/mongock/default/kpi_master.json [hunk 4]
```diff
@@ -4020,6 +4020,47 @@
...
+            "text": "Detailed Information at",
+            "link": "https://psknowhow.atlassian.net/wiki/spaces/PSKNOWHOW/pages/92930049/Build+Frequency"
+          }
+        }
+      ]
+    },
+    "xAxisLabel": "Weeks",
+    "yAxisLabel": "Builds Count",
+    "isPositiveTrend": false,
+    "showTrend": true,
```
Comment (shunaray) [inline customapi/src/main/resources/json/mongock/default/kpi_master.json:4051]: -ve positiveTrend kpi ?

File: customapi/src/main/resources/json/mongock/default/kpi_category_mapping.json [hunk 5]
```diff
@@ -255,5 +255,11 @@
     "categoryId": "categoryTwo",
     "kpiOrder": 15,
     "kanban": false
+  },
+  {
```
Comment (shunaray) [inline customapi/src/main/resources/json/mongock/default/kpi_category_mapping.json:259]: for internal client speed dashboard,

File: customapi/src/main/java/com/publicissapient/kpidashboard/apis/util/KPIExcelUtility.java [hunk 6]
```diff
@@ -957,6 +958,30 @@ public static void populateCodeBuildTime(List<KPIExcelData> kpiExcelData, String
...
+
+		for (int i = 0; i < buildFrequencyInfo.getBuildJobList().size(); i++) {
+			KPIExcelData excelData = new KPIExcelData();
+			excelData.setProjectName(projectName);
+			excelData.setJobName(buildFrequencyInfo.getBuildJobList().get(i));
+			Map<String, String> buildUrl = new HashMap<>();
+			buildUrl.put(buildFrequencyInfo.getBuildUrlList().get(i), buildFrequencyInfo.getBuildUrlList().get(i));
+			excelData.setBuildUrl(buildUrl);
+			excelData.setStartDate(buildFrequencyInfo.getBuildStartTimeList().get(i));
+			excelData.setWeeks(buildFrequencyInfo.getWeeksList().get(i));
```
Comment (shunaray) [inline customapi/src/main/java/com/publicissapient/kpidashboard/apis/util/KPIExcelUtility.java:980]: check once date format DD/MMM/YY

File: customapi/src/main/java/com/publicissapient/kpidashboard/apis/jenkins/service/BuildFrequencyServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,353 @@
...
+	}
+
+	/**
+	 * 
+	 * @param monday
+	 * @param sunday
+	 * @param build
+	 * @return
+	 */
+	private boolean checkDateIsInWeeks(LocalDate monday, LocalDate sunday, Build build) {
```
Comment (shunaray) [inline customapi/src/main/java/com/publicissapient/kpidashboard/apis/jenkins/service/BuildFrequencyServiceImpl.java:338]: use dateUtils method

File: customapi/src/main/java/com/publicissapient/kpidashboard/apis/jenkins/service/BuildFrequencyServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,353 @@
...
+
+		return kpiElement;
+	}
+
+	@Override
+	public Long calculateKPIMetrics(Map<ObjectId, List<Build>> builds) {
+		return null;
+	}
+
+	@Override
```
Comment (shunaray) [inline customapi/src/main/java/com/publicissapient/kpidashboard/apis/jenkins/service/BuildFrequencyServiceImpl.java:135]: add missing javadoc

File: customapi/src/main/resources/json/mongock/default/kpi_master.json (hunk 3 above)
Comment (aksshriv1) [reply customapi/src/main/resources/json/mongock/default/kpi_master.json:4034]: done

**Review by shunaray — APPROVED** [review]

**Pull request merged**